README.md
# ApplozicSwift

[![Version](https://img.shields.io/cocoapods/v/ApplozicSwift.svg?style=flat)](http://cocoapods.org/pods/ApplozicSwift)
![iOS 8.0+](https://img.shields.io/badge/iOS-9.0%2B-blue.svg)
![Xcode 8.2+](https://img.shields.io/badge/Xcode-8.2%2B-blue.svg)
![Swift 3.0+](https://img.shields.io/badge/Swift-3.0%2B-orange.svg)
[![License](https://img.shields.io/cocoapods/l/Material.svg?style=flat)](https://github.com/lkzhao/Material/blob/master/LICENSE?raw=true)

UI kit for Applozic SDK, written completely in Swift.

## Overview

Open source iOS Chat and Messaging SDK that lets you add real time messaging in your mobile (android, iOS) applications and website.

Signup at https://www.applozic.com/signup.html to get the application key.

------------------------

![Screenshot0][img0] &nbsp;&nbsp; ![Screenshot1][img1] &nbsp;&nbsp;

![Screenshot2][img2] &nbsp;&nbsp;

## Sample Project

There's a sample project in the Demo directory. To use it, run `pod install` to download the required libraries. Have fun!

## Project Status

This project is actively under development.

## Installation

### CocoaPods

For ApplozicSwift, use the following entry in your Podfile:

`pod 'ApplozicSwift'`

Then run `pod install`.

In any file you'd like to use ApplozicSwift in, don't forget to
import the framework with `import ApplozicSwift`.

### Manually

- Open up Terminal, `cd` into your top-level project directory, and run the following command *if* your project is not initialized as a git repository:

```bash
$ git init
```

- Add ApplozicSwift, Applozic, Kingfisher & MGSwipeTableCell as a git [submodule](http://git-scm.com/docs/git-submodule) by running the following commands:

```bash
$ git submodule add https://github.com/AppLozic/ApplozicSwift
$ git submodule add https://github.com/AppLozic/Applozic-iOS-SDK
$ git submodule add https://github.com/onevcat/Kingfisher.git
$ git submodule add https://github.com/MortimerGoro/MGSwipeTableCell.git
```

- Open the new `ApplozicSwift` folder, and drag the `ApplozicSwift.xcodeproj` into the Project Navigator of your application's Xcode project. Do the same with the `Applozic.xcodeproj` in the `Applozic` folder, `Kingfisher.xcodeproj` in the `Kingfisher` folder and `MGSwipeTableCell.xcodeproj` in the `MGSwipeTableCell` folder.

> They should appear nested underneath your application's blue project icon. Whether it is above or below all the other Xcode groups does not matter.

- Verify that the deployment targets of the `xcodeproj`s match that of your application target in the Project Navigator.
- Next, select your application project in the Project Navigator (blue project icon) to navigate to the target configuration window and select the application target under the "Targets" heading in the sidebar.
- In the tab bar at the top of that window, open the "General" panel.
- Click on the `+` button under "Embedded Binaries" again and add the build target you need for `ApplozicSwift`.
- Click on the `+` button under "Embedded Binaries" again and add the build target you need for `Applozic`.
- Click on the `+` button under "Embedded Binaries" again and add the build target you need for `KingFisher`.
- Click on the `+` button again and add the correct build target for `MGSwipeTableCell`.

- And that's it!

> The four frameworks are automagically added as a target dependency, linked framework and embedded framework in a copy files build phase which is all you need to build on the simulator and a device.

## Usage

#### ChatManager

Add the [ALChatManager](https://github.com/AppLozic/ApplozicSwift/blob/master/Demo/ApplozicSwiftDemo/ALChatManager.swift) file in your project. It contains the convenient methods required to handle the chat like registration, opening chat list etc.

#### Registration

```swift
// Use the same API for login
let alChatManager = ALChatManager(applicationKey: ALChatManager.applicationId as NSString)
alChatManager.registerUser(alUser, completion: {response, error in
      if error == nil {
          NSLog("[REGISTRATION] Applozic user registration was successful: %@ \(response?.isRegisteredSuccessfully())")
          // Launch chat
      } else {
          NSLog("[REGISTRATION] Applozic user registration error: %@", error.debugDescription)
      }
})
```

#### Launch chat list

```swift
let alChatManager = ALChatManager(applicationKey: <Application key>, with: ALKConfiguration())
alChatManager.launchChatList(from: self)
```

#### Launch Group

```swift
let alChatManager = ALChatManager(applicationKey: <Application key>, configuration: ALKConfiguration())
alChatManager.launchGroupWith(clientGroupId: <groupId>, from: self)
```


#### Launch Individual Chat

```swift
let alChatManager = ALChatManager(applicationKey: <Application key>, configuration: ALKConfiguration())
alChatManager.launchChatWith(contactId: <userId>, from: self)
```


## Requirements

* iOS 9.0+
* Xcode 10.0+
* Swift versions :: 
3.0 - Use SDK version below 1.1.0
4.0 - Use SDK version starting from 1.1.0

## Contributing

We would love you for the contribution to ApplozicSwift, check the LICENSE file for more info.


## License

ApplozicSwift is released under a BSD 3-Clause. See [LICENSE](LICENSE) for more information.

[img0]:https://raw.githubusercontent.com/Applozic/ApplozicSwift/master/Screenshots/screenshot0.png
[img1]:https://raw.githubusercontent.com/Applozic/ApplozicSwift/master/Screenshots/screenshot1.png
[img2]:https://raw.githubusercontent.com/Applozic/ApplozicSwift/master/Screenshots/screenshot2.png


CHANGELOG.md
# CHANGELOG

The changelog for [ApplozicSwift](https://github.com/AppLozic/ApplozicSwift). Also see the [releases](https://github.com/AppLozic/ApplozicSwift/releases) on Github.


## [Unreleased]

### Enhancements
- [CM-380] Support HTML message type with rich messages.
- [CM-213] Added support for sharing photos saved on iCloud.

## [5.9.2] - 2020-08-21

### Enhancements
- Removed the line limit of form selection items to allow dynamic text of any size.

## [5.9.1] - 2020-08-18

### Enhancements
- [CM-426] Increased the line limit of form selection items.

### Fixes
- Fixed issue where message metadata coming as nil in some cases.

## [5.9.0] - 2020-08-14

### Enhancements
- [CM-384] Added document sharing support.
### Fixes
- [CM-417] Fixed a crash that occurred in iOS 13.5 after going back from the group detail or document viewer to a conversation thread.

## [5.8.2] - 2020-07-24

### Fixes
- Fixed the issue in attachments download.

## [5.8.1] - 2020-07-23

### Fixes
- Fixed the issue where form data submit was using data from previous cell.

## [5.8.0] - 2020-07-15

### Enhancements
- [CM-22] Add form template support in the rich messages.

## [5.7.1] - 2020-07-06

### Fixes
- [CM-354] Fixed audio playing issue when changing chat threads.

## [5.7.0] - 2020-06-23

### Enhancements
- Migrated to Swift 5

### Fixes
- [CM-281] Fixed "no results found" error when opening a location message.

## [5.6.0] - 2020-06-08

### Fixes
[CM-276] Fixed a crash in MQTT subscription

## [5.5.0] - 2020-05-25

### Enhancements
-[CM-224] Add support for rich message templates without any text message

## [5.4.0] - 2020-05-07

### Enhancements

- Show link preview in chat for web links
</br>You can use below config to disable web link preview in chat:
```
    config.isLinkPreviewDisabled = true
```
You need to set `Allow Arbitrary` Loads to YES on your project's `Info.plist` file.
```
 <key>NSAppTransportSecurity</key>
  <dict>
      <key>NSAllowsArbitraryLoads</key>
      <true/>
  </dict>
```
If you don't want to use the above option, You can white list some the websites which you want to allow them in `Info.plist` file
```
<key>NSAppTransportSecurity</key>
<dict>
  <key>NSAllowsArbitraryLoads</key>
  <false/>
  <key>NSExceptionDomains</key>
  <dict>
    <key>google.com</key>
    <dict>
      <key>NSIncludesSubdomains</key>
      <true/>
      <key>NSExceptionAllowsInsecureHTTPLoads</key>
      <true/>
    </dict>
  </dict>
</dict>
```
You can read more on Property List Keys
[here](https://developer.apple.com/documentation/bundleresources/information_property_list/nsapptransportsecurity/nsexceptiondomains)

### Fixes
- [CM-146] Fixed back button issue in searched conversation view.
- [CM-241] Fixed a crash that could happen when some message updates are delayed.
- [CM-225] Fixed an issue where group profile was getting removed on group name update.

## [5.3.0] - 2020-04-21

### Enhancements

- [CM-193] Added support to open the application when tapping any URL in the text message. If the application isn't installed, we'll open it in Safari.

### Fixes

- [CM-201] Fixed an issue where the footer of the sender-side view of email messages overlapped with the header, as there were no constraints.
- [CM-146] Fixed search messages loading issue.

## [5.2.0] - 2020-03-31

### Enhancements
- Customize the date and info message font, text color or background color of the view
</br>Use the below config to change the style of date separator and channel info messages:

```
   ALKMessageStyle.infoMessage = Style(font: UIFont.systemFont(ofSize: 12), text: UIColor.black , background: .red)

   ALKMessageStyle.dateSeparator = Style(font: UIFont.systemFont(ofSize: 12), text: UIColor.black, background: .red)
```

## [5.1.1] - 2020-03-04

### Fixes
-[CM-136] RichMessage button UI issue

## [5.1.0] - 2020-03-03

### Enhancements
- Added support for syncing package details when a suspension screen is shown.
- Added an option to change message status(read receipt) style. Now icon for message status can be changed from outside or it can be hidden.
- Added a config to disable in-app notification banner.
</br>You can use below config to disable in-app notification banner:
```
	config.isInAppNotificationBannerDisabled = true
```

## [5.0.0] - 2020-02-14

### Enhancements
- Added save photo button in the media viewer.

### Fixes
- If the keyboard used is not English, the keyboard is changed. (.asciiCapable -> .emailAddress)
- [CM-141] Fixed an issue where messages were deleted after leaving the group.
- [CM-142] Fixed back button alignment issue in the Conversation screen.
- [CM-133] Fixed an issue where updating certain localizable strings from a custom localizable file did not work.

## [4.1.0] - 2020-01-27

### Enhancements

- [CM-51] Email message layout improvement.

## [4.0.1]
---
### Fixes

- Fixed an issue where button's title was sent as a message in card template.

## [4.0.0]
---
### Enhancements
- [AL-3826] Reply message will be shown in a popup when it is not visible in the list.
- Now user's display name will be shown instead of group name in case of `GROUP_OF_TWO`.

### Fixes
- Fixed an issue where navigation bar's back button color was not changing using `UIAppearance`.
- Fixed all the warnings.

## [3.4.0]
---
### Enhancements
- Added support for iOS 13.
- [AL-3761] Added support for member mention in groups.
- [AL-3741] Added support for rich message template 11.
- [AL-3847] Updated UI for link button and refractored rich messages to use same view for all types of buttons.
- [AL-3642] Added support for message search.

Sources/Views/ALKChatCell.swift
//
//  ChatCell.swift
//  
//
//  Created by Mukesh Thawani on 04/05/17.
//  Copyright © 2017 Applozic. All rights reserved.
//

import Foundation
import UIKit
import Kingfisher
import MGSwipeTableCell
import Applozic

protocol ALKChatViewModelProtocol {
    var avatar: URL? { get }
    var avatarImage: UIImage? { get }
    var avatarGroupImageUrl: String? { get }
    var name: String { get }
    var groupName: String { get }
    var theLastMessage: String? { get }
    var hasUnreadMessages: Bool { get }
    var totalNumberOfUnreadMessages: UInt { get }
    var isGroupChat: Bool { get }
    var contactId: String? { get }
    var channelKey: NSNumber? { get }
    var conversationId: NSNumber! {get set}
    var createdAt: String? { get }
}

enum ALKChatCellAction {
    case delete
    case favorite
    case store
    case call
    case mute
}

protocol ALKChatCellDelegate: class {
    func chatCell(cell: ALKChatCell, action: ALKChatCellAction, viewModel: ALKChatViewModelProtocol)
}

final class ALKChatCell: MGSwipeTableCell {

    private var avatarImageView: UIImageView = {
        let imv = UIImageView()
        imv.contentMode = .scaleAspectFill
        imv.clipsToBounds = true
        let layer = imv.layer
        layer.cornerRadius = 22.5
        layer.backgroundColor = UIColor.clear.cgColor
        layer.masksToBounds = true
        return imv
    }()

    private var nameLabel: UILabel = {
        let label = UILabel()
        label.lineBreakMode = .byTruncatingTail
        label.numberOfLines = 1
        label.font = Font.bold(size: 14.0).font()
        label.textColor = .text(.black00)
        return label
    }()

    private var locationLabel: UILabel = {
        let label = UILabel()
        label.lineBreakMode = .byTruncatingTail
        label.numberOfLines = 1
        label.font = Font.normal(size: 14.0).font()
        label.textColor = UIColor(netHex: 0x9B9B9B)
        return label
    }()

    private var lineView: UIView = {
        let view = UIView()
        let layer = view.layer
        view.backgroundColor = UIColor.init(red: 200.0/255.0, green: 199.0/255.0, blue: 204.0/255.0, alpha: 0.33)
        return view
    }()

    private lazy var voipButton: UIButton = {
        let bt = UIButton(type: .custom)
        bt.setImage(UIImage(named: "icon_menu_dial_on"), for: .normal)
        bt.setImage(UIImage(named: "icon_call_disable"), for: .disabled)
        bt.addTarget(self, action: #selector(callTapped(button:)), for: .touchUpInside)
        return bt
    }()

    private lazy var favoriteButton: UIButton = {
        let bt = UIButton(type: .custom)
        bt.setImage(UIImage(named: "icon_favorite"), for: .normal)
        bt.setImage(UIImage(named: "icon_favorite_active"), for: .highlighted)
        bt.setImage(UIImage(named: "icon_favorite_active"), for: .selected)
        bt.addTarget(self, action: #selector(favoriteTapped(button:)), for: UIControl.Event.touchUpInside)
        return bt
    }()

    // MARK: BadgeNumber
    private lazy var badgeNumberView: UIView = {
        let view = UIView(frame: .zero)
        view.setBackgroundColor(color: .main)
        return view
    }()

    private lazy var badgeNumberLabel: UILabel = {
        let label = UILabel(frame: .zero)
        label.text = "0"
        label.textAlignment = .center
        label.setTextColor(color: .white)
        label.setFont(font: .normal(size: 9.0))

        return label
    }()

    private var timeLabel: UILabel = {
        let label = UILabel()
        label.textAlignment = .right
        label.numberOfLines = 1
        label.font = Font.normal(size: 14.0).font()
        label.textColor = UIColor(netHex: 0x9B9B9B)
        return label
    }()

    private var onlineStatusView: UIView = {
        let view = UIView(frame: .zero)
        view.backgroundColor = UIColor.onlineGreen()
        return view
    }()
    
    private var checkBox: UIView = {
        let view = UIView()
        view.layer.borderColor = UIColor.darkGray.cgColor
        view.layer.cornerRadius = 5
        view.layer.borderWidth = 2
        return view
    }()

    private var avatarName: UILabel = {
        let label = UILabel()
        label.backgroundColor = UIColor.gray
        label.layer.cornerRadius = 22.5
        label.textColor = UIColor.white
        label.textAlignment = .center
        label.numberOfLines = 1
        label.clipsToBounds = true
        label.font = Font.bold(size: 20.0).font()
        return label
    }()

    weak var chatCellDelegate: ALKChatCellDelegate?

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        selectionStyle = .none
        voipButton.isHidden = true
        disableMultiSelect()
        setupConstraints()
    }

    deinit {
        voipButton.removeTarget(self, action:  #selector(callTapped(button:)), for: .touchUpInside)
        //favoriteButton.removeTarget(self, action:  #selector(favoriteTapped(button:)), for: .touchUpInside)
    }

    override func setHighlighted(_ highlighted: Bool, animated: Bool) {
        super.setHighlighted(highlighted, animated: animated)

        guard let viewModel = viewModel else {
            return
        }

        lineView.backgroundColor = UIColor(netHex: 0xF1F1F1)

        backgroundColor = highlighted ? UIColor.init(netHex: 0xECECEC) : UIColor.white
        contentView.backgroundColor = backgroundColor

        // set backgroundColor of badgeNumber
        badgeNumberView.setBackgroundColor(color: .main)
    }
    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        guard let viewModel = viewModel else {
            return
        }


        lineView.backgroundColor = UIColor(netHex: 0xF1F1F1)

        backgroundColor = selected ? UIColor.init(netHex: 0xECECEC) : UIColor.white
        contentView.backgroundColor = backgroundColor

        // set backgroundColor of badgeNumber
        badgeNumberView.setBackgroundColor(color: .main)
    }

    var viewModel: ALKChatViewModelProtocol?

    func update(viewModel: ALKChatViewModelProtocol, identity: ALKIdentityProtocol?) {

        self.viewModel = viewModel
        let placeHolder = UIImage(named: "placeholder", in: Bundle.applozic, compatibleWith: nil)
        avatarImageView.isHidden = true
        avatarName.isHidden = true
        avatarName.backgroundColor = UIColor.gray

        if let avatarImage = viewModel.avatarImage {
            if let imgStr = viewModel.avatarGroupImageUrl,let imgURL = URL.init(string: imgStr) {
                avatarImageView.isHidden = false
                let resource = ImageResource(downloadURL: imgURL, cacheKey: imgStr)
                avatarImageView.kf.setImage(with: resource, placeholder: avatarImage, options: nil, progressBlock: nil, completionHandler: nil)
            } else {
                avatarName.isHidden = false
                let name = getFirstTwoLetters(text: viewModel.groupName)
                avatarName.text = name.isEmpty ? "X": name
            }

        }else if let avatar = viewModel.avatar {
            avatarImageView.isHidden = false
            let resource = ImageResource(downloadURL: avatar, cacheKey: avatar.absoluteString)
            avatarImageView.kf.setImage(with: resource, placeholder: placeHolder, options: nil, progressBlock: nil, completionHandler: nil)
        } else {
            avatarName.isHidden = false
            let name = getFirstTwoLetters(text: viewModel.name)
            avatarName.text = name.isEmpty ? "X": name
        }

        let name = viewModel.isGroupChat ? viewModel.groupName:viewModel.name
        nameLabel.text = name
        locationLabel.text = viewModel.theLastMessage

        let deleteButton = MGSwipeButton.init(type: .system)
        deleteButton.backgroundColor = UIColor.mainRed()
        deleteButton.setImage(UIImage(named: "icon_delete_white", in: Bundle.applozic, compatibleWith: nil), for: .normal)
        deleteButton.setImage(UIImage(named: "icon_delete_white", in: Bundle.applozic, compatibleWith: nil), for: .highlighted)
        deleteButton.setImage(UIImage(named: "icon_delete_white", in: Bundle.applozic, compatibleWith: nil), for: .selected)
        deleteButton.tintColor = .white
        deleteButton.frame = CGRect.init(x: 0, y: 0, width: 69, height: 69)
        deleteButton.callback = { [weak self] (buttnon) in

            guard let strongSelf = self else {return true}
            guard let viewModel = strongSelf.viewModel else {return true}

            strongSelf.chatCellDelegate?.chatCell(cell: strongSelf, action: .delete, viewModel: viewModel)

            return true
        }

        let muteButton = MGSwipeButton.init(type: .custom)
        muteButton.backgroundColor = UIColor.init(netHex: 0x999999)
        muteButton.setImage(UIImage(named: "icon_mute_inactive", in: Bundle.applozic, compatibleWith: nil), for: .normal)
        muteButton.setImage(UIImage(named: "icon_mute_active", in: Bundle.applozic, compatibleWith: nil), for: .selected)
        muteButton.frame = CGRect.init(x: 0, y: 0, width: 69, height: 69)
        muteButton.callback = { [weak self] (buttnon) in

            guard let strongSelf = self else {return true}
            guard let viewModel = strongSelf.viewModel else {return true}

            strongSelf.chatCellDelegate?.chatCell(cell: strongSelf, action: .mute, viewModel: viewModel)

            return true
        }

        self.rightButtons = [muteButton]
        self.rightSwipeSettings.transition = .static

        self.leftButtons = [deleteButton]
        self.leftSwipeSettings.transition = .static

        // get unread count of message and set badgenumber
        let unreadMsgCount = viewModel.totalNumberOfUnreadMessages
        let numberText: String = (unreadMsgCount < 1000 ? "\(unreadMsgCount)" : "999+")
        let isHidden = (unreadMsgCount < 1)

        badgeNumberView.isHidden = isHidden
        badgeNumberLabel.text = numberText

        timeLabel.text = viewModel.createdAt
        onlineStatusView.isHidden = true

        if !viewModel.isGroupChat {
            
            let contactService = ALContactService()
            guard let contactId = viewModel.contactId,
            let contact = contactService.loadContact(byKey: "userId", value: contactId) else {
                return
            }

            if contact.block || contact.blockBy || contactService.isUserDeleted(contactId) {
                onlineStatusView.isHidden = true
                return
            }

            onlineStatusView.isHidden = !contact.connected
        }

        self.voipButton.isEnabled = !viewModel.isGroupChat
    }

    private func setupConstraints() {

        contentView.addViewsForAutolayout(views: [avatarImageView, nameLabel, locationLabel,lineView,voipButton,/*favoriteButton,*/avatarName,badgeNumberView, timeLabel, checkBox, onlineStatusView])

        // setup constraint of imageProfile
        avatarImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 17.0).isActive = true
        avatarImageView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 15.0).isActive = true
        avatarImageView.heightAnchor.constraint(equalToConstant: 45.0).isActive = true
        avatarImageView.widthAnchor.constraint(equalToConstant: 45.0).isActive = true

        // setup constraint of name
        nameLabel.topAnchor.constraint(equalTo: avatarImageView.topAnchor, constant: 2).isActive = true
        nameLabel.heightAnchor.constraint(equalToConstant: 20).isActive = true
        nameLabel.leadingAnchor.constraint(equalTo: avatarImageView.trailingAnchor, constant: 12).isActive = true
        nameLabel.trailingAnchor.constraint(equalTo: timeLabel.leadingAnchor, constant: -5).isActive = true

        // setup constraint of mood
        locationLabel.topAnchor.constraint(equalTo: nameLabel.bottomAnchor, constant: 2).isActive = true
        locationLabel.heightAnchor.constraint(equalToConstant: 20).isActive = true
        locationLabel.leadingAnchor.constraint(equalTo: avatarImageView.trailingAnchor, constant: 12).isActive = true
        locationLabel.trailingAnchor.constraint(equalTo: checkBox.leadingAnchor, constant: -19).isActive = true

        // setup constraint of line
        lineView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor).isActive = true
        lineView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor).isActive = true
        lineView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor).isActive = true
        lineView.heightAnchor.constraint(equalToConstant: 1).isActive = true

        // setup constraint of favorite button
        /*
         favoriteButton.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -15).isActive = true
         favoriteButton.centerYAnchor.constraint(equalTo: contentView.centerYAnchor).isActive = true
         favoriteButton.widthAnchor.constraint(equalToConstant: 24.0).isActive = true
         favoriteButton.heightAnchor.constraint(equalToConstant: 24.0).isActive = true
         */

        // setup constraint of VOIP button
        //voipButton.trailingAnchor.constraint(equalTo: favoriteButton.leadingAnchor, constant: -25.0).isActive = true
        voipButton.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -23).isActive = true
        voipButton.centerYAnchor.constraint(equalTo: contentView.centerYAnchor).isActive = true
        voipButton.widthAnchor.constraint(equalToConstant: 24.0).isActive = true
        voipButton.heightAnchor.constraint(equalToConstant: 25.0).isActive = true

        // setup constraint of badgeNumber
        badgeNumberView.addViewsForAutolayout(views: [badgeNumberLabel])


        badgeNumberView.trailingAnchor.constraint(lessThanOrEqualTo: nameLabel.leadingAnchor, constant: -5)
        badgeNumberView.topAnchor.constraint(equalTo: avatarImageView.topAnchor, constant: 0).isActive = true
        badgeNumberView.leadingAnchor.constraint(equalTo: avatarImageView.trailingAnchor, constant: -12).isActive = true

        badgeNumberLabel.setContentCompressionResistancePriority(UILayoutPriority(rawValue: 1000), for: .horizontal)
        badgeNumberLabel.topAnchor.constraint(equalTo: badgeNumberView.topAnchor, constant: 2.0).isActive = true
        badgeNumberLabel.bottomAnchor.constraint(equalTo: badgeNumberView.bottomAnchor, constant: -2.0).isActive = true
        badgeNumberLabel.leadingAnchor.constraint(equalTo: badgeNumberView.leadingAnchor, constant: 2.0).isActive = true
        badgeNumberLabel.trailingAnchor.constraint(equalTo: badgeNumberView.trailingAnchor, constant: -2.0).isActive = true
        badgeNumberLabel.widthAnchor.constraint(greaterThanOrEqualToConstant: 11.0).isActive = true
        badgeNumberLabel.heightAnchor.constraint(greaterThanOrEqualToConstant: 11.0).isActive = true

        timeLabel.heightAnchor.constraint(equalToConstant: 15).isActive = true
        timeLabel.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -19).isActive = true
        timeLabel.widthAnchor.constraint(equalToConstant: 70).isActive = true
        timeLabel.topAnchor.constraint(equalTo: nameLabel.topAnchor, constant: 0).isActive  = true

        onlineStatusView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 0).isActive = true
        onlineStatusView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: 0).isActive = true
        onlineStatusView.widthAnchor.constraint(equalToConstant: 6).isActive = true

        avatarName.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 17.0).isActive = true
        avatarName.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 15.0).isActive = true
        avatarName.heightAnchor.constraint(equalToConstant: 45.0).isActive = true
        avatarName.widthAnchor.constraint(equalToConstant: 45.0).isActive = true
        
        checkBox.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -25).isActive = true
        checkBox.centerYAnchor.constraint(equalTo: contentView.centerYAnchor).isActive = true
        checkBox.widthAnchor.constraint(equalToConstant: 25.0).isActive = true
        checkBox.heightAnchor.constraint(equalToConstant: 25.0).isActive = true

        // update frame
        contentView.layoutIfNeeded()

        badgeNumberView.layer.cornerRadius = badgeNumberView.frame.size.height / 2.0
    }


    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private weak var comingSoonDelegate: UIView?

    func setComingSoonDelegate(delegate: UIView) {
        comingSoonDelegate = delegate
    }

    @objc func callTapped(button: UIButton) {

        guard let viewModel = self.viewModel else {return}
        self.chatCellDelegate?.chatCell(cell: self, action: .call, viewModel:viewModel)
    }

    @objc func favoriteTapped(button: UIButton) {
//        comingSoonDelegate?.makeToast(SystemMessage.ComingSoon.Favorite, duration: 1.0, position: .center)
    }

    override func setEditing(_ editing: Bool, animated: Bool) {
        super.setEditing(editing, animated: animated)
    }

    private func getFirstTwoLetters(text: String) -> String {
        let stringInputArr = text.components(separatedBy: " ")
        var firstTwoCharStr = ""

        for string in stringInputArr {
            guard let firstChar = string.characters.first else { continue }
            firstTwoCharStr = firstTwoCharStr + String(firstChar)
        }
        return firstTwoCharStr
    }

    private func getRandomColor() -> UIColor {
        let colors = [0x19A5E4, 0x0EB04B, 0xF3B618, 0xE4E9EC]
        let x = randomInt(min: 0, max: 3)
        return UIColor.init(netHex: colors[x])
    }

    func randomInt(min: Int, max:Int) -> Int {
        return min + Int(arc4random_uniform(UInt32(max - min + 1)))
    }
    
    func enableMultiSelect() {
        checkBox.isHidden = false
        timeLabel.isHidden = true
    }
    
    func disableMultiSelect() {
        checkBox.isHidden = true
        timeLabel.isHidden = false
    }
    
    func selectCheckBox() {
        //checkMark Image as UIView background
        UIGraphicsBeginImageContext(checkBox.frame.size)
        let img = UIImage(named: "checkboxBackground", in: Bundle.applozic, compatibleWith: nil)
        img?.drawAsPattern(in: checkBox.bounds)
        if let image = UIGraphicsGetImageFromCurrentImageContext(){
            UIGraphicsEndImageContext()
            checkBox.backgroundColor = UIColor(patternImage: image)
        }else{
            UIGraphicsEndImageContext()
            debugPrint("Image not available")
            checkBox.backgroundColor = ALKConfiguration.init().customPrimary
        }
    }
    
    func deselectCheckBox() {
        checkBox.backgroundColor = UIColor.clear
    }
}
